lambdas/bearer-token-handler/src/bearer-token-handler.ts

```diff
@@ -2,54 +2,123 @@ import { LambdaInterface } from "@aws-lambda-powertools/commons";
+        body: this.createRequestBody(totpCode.totp, clientSecret, clientId),
       });
 
-      if (response.ok) {
-        const body = (await response.json()) as OAuthResponse;
-        const expiry = body.expires_in / 2;
-        return {
-          token: body.access_token,
-          tokenExpiry: (Date.now() + expiry * 1000).toString(),
-          tokenExpiryInMinutes: expiry / 60,
-        };
+      return await this.handleResponse(response);
+    } catch (error: unknown) {
+      if (error instanceof Error) {
+        logger.error("Error in BearerTokenHandler: " + error.message);
       }
+      throw error;
+    }
+  }
 
-      throw new Error(
-        `Error response received from HMRC ${response.status} ${response.statusText}`
+  private async getSecrets(
+    stackName: string,
+    tokenType: string
+  ): Promise<[string, string, string]> {
+    const totpSecret = await this.getSecret(
+      format(totpSecretName, stackName, tokenType)
+    );
+    const clientId = await this.getSecret(
+      format(clientIdName, stackName, tokenType)
+    );
+    const clientSecret = await this.getSecret(
+      format(clientSecretName, stackName, tokenType)
+    );
+    return [totpSecret, clientId, clientSecret];
+  }
+
+  private createRequestBody(
+    totpCode: string,
+    clientSecret: string,
+    clientId: string
+  ): string {
+    const data = {
+      client_secret: totpCode + clientSecret,
+      client_id: clientId,
+      grant_type: "client_credentials",
+    } as Record<string, string>;
+    return Object.keys(data)
+      .map(
+        (key) => encodeURIComponent(key) + "=" + encodeURIComponent(data[key])
+      )
+      .join("&");
+  }
+
+  private async handleResponse(response: Response): Promise<unknown> {
+    if (response.ok) {
+      const body = (await response.json()) as OAuthResponse;
+      const expiry = body.expires_in / 2;
+      return {
+        token: body.access_token,
+        tokenExpiry: (Date.now() + expiry * 1000).toString(),
+        tokenExpiryInMinutes: expiry / 60,
+      };
+    }
+    throw new Error(
+      `Error response received from HMRC ${response.status} ${response.statusText}`
+    );
+  }
+
+  private async getSecret(secretName: string) {
+    try {
+      const data = await this.secretsManager.send(
+        new GetSecretValueCommand({
+          SecretId: secretName,
+        })
       );
+      if (!data.SecretString) {
+        throw new Error("No secret found for " + secretName);
+      }
+      return data.SecretString;
     } catch (error: unknown) {
       if (error instanceof Error) {
-        logger.error("Error in BearerTokenHandler: " + error.message);
+        logger.error(
+          "Failed to retrieve secret " + secretName + ":" + error.message
+        );
       }
       throw error;
     }
   }
 }
 
-const handlerClass = new BearerTokenHandler();
+function format(str: string, ...args: string[]): string {
+  return str.replace(/%s/g, () => args.shift() ?? "");
```

Comment:
can this be simplified with string laterial `const clientSecretName = `HMRC/ClientSecret/{stackName}/{tokenType}`